### PR TITLE
PWX-32147 : add custom image registry support for plugin pod

### PR DESCRIPTION
**Pull request by nikita-bhatia**
Added custom Image Registery for px-plugin and px-plugin-proxy pod : 
1. Added DynamicPlugin and Nginxc image in manifest and Desired images
2. Changed deployment creation to use desired image at time of creation

Tests done : 
Tested on air-gapped cluster with CIR "docker.pwx.dev.purestorage.com" , both px-plugin and px-plugin-proxy pods were up and running fine

image

Events:
  Type     Reason          Age    From               Message
  ----     ------          ----   ----               -------
  Normal   Scheduled       7m21s  default-scheduler  Successfully assigned kube-system/px-plugin-proxy-d5c6f65d7-vkfqr to airgap-ocp-w8pqc-worker-2bl5l by airgap-ocp-w8pqc-master-0
  Warning  FailedMount     7m21s  kubelet            MountVolume.SetUp failed for volume "nginx-certs" : secret "px-plugin-proxy" not found
  Normal   AddedInterface  7m19s  multus             Add eth0 [10.129.2.23/23] from openshift-sdn
  Normal   Pulling         7m19s  kubelet            Pulling image "docker.pwx.dev.purestorage.com/nginxinc/nginx-unprivileged:1.23"
  Normal   Pulled          7m15s  kubelet            Successfully pulled image "docker.pwx.dev.purestorage.com/nginxinc/nginx-unprivileged:1.23" in 3.518983541s
  Normal   Created         7m15s  kubelet            Created container nginx
  Normal   Started         7m15s  kubelet            Started container nginx

Events:
  Type    Reason          Age    From               Message
  ----    ------          ----   ----               -------
  Normal  Scheduled       7m42s  default-scheduler  Successfully assigned kube-system/px-plugin-76bcf74b46-9jxks to airgap-ocp-w8pqc-worker-q9ljd by airgap-ocp-w8pqc-master-0
  Normal  AddedInterface  7m40s  multus             Add eth0 [10.130.2.74/23] from openshift-sdn
  Normal  Pulling         7m40s  kubelet            Pulling image "docker.pwx.dev.purestorage.com/portworx/portworx-dynamic-plugin:1.0.0"
  Normal  Pulled          7m37s  kubelet            Successfully pulled image "docker.pwx.dev.purestorage.com/portworx/portworx-dynamic-plugin:1.0.0" in 2.68198716s
  Normal  Created         7m37s  kubelet            Created container px-plugin
  Normal  Started         7m37s  kubelet            Started container px-plugin

### CODE REVIEW

File: drivers/storage/portworx/component/plugin.go [hunk 1]
```diff
@@ -354,3 +360,23 @@ func isVersionSupported(v string) bool {
 
 	return currentVersion.GreaterThanOrEqual(targetVersion)
 }
+
+func getDesiredPluginImage(cluster *corev1.StorageCluster) string {
+	imageName := "portworx/portworx-dynamic-plugin:1.0.0"
```
Comment (zoxpx) [inline drivers/storage/portworx/component/plugin.go:365]: Let's please move the `portworx/portworx-dynamic-plugin:1.0.0` as one of the constants -- ideally together w/ other image-constants in drivers/storage/portworx/manifest/manifest.go ?

File: drivers/storage/portworx/component/plugin.go [hunk 2]
```diff
@@ -354,3 +360,23 @@ func isVersionSupported(v string) bool {
...
+
+	if cluster.Status.DesiredImages != nil && cluster.Status.DesiredImages.DynamicPlugin != "" {
+		imageName = cluster.Status.DesiredImages.DynamicPlugin
+	}
+	imageName = util.GetImageURN(cluster, imageName)
+	return imageName
+}
+
+func getDesiredNginxcImage(cluster *corev1.StorageCluster) string {
+	imageName := "nginxinc/nginx-unprivileged:1.23"
```
Comment (zoxpx) [inline drivers/storage/portworx/component/plugin.go:375]: Same as before -- let's declare it as `const` and (if possible) move to `manifest.go`

**Review by zoxpx — APPROVED** [review]
I think this looks OK -- a couple of minor issues listed below

File: drivers/storage/portworx/manifest/manifest.go [hunk 3]
```diff
@@ -85,6 +85,8 @@ type Release struct {
 	KubeScheduler              string `yaml:"kubeScheduler,omitempty"`
 	KubeControllerManager      string `yaml:"kubeControllerManager,omitempty"`
 	Pause                      string `yaml:"pause,omitempty"`
+	DynamicPlugin              string `yaml:"dynamicPlugin,omitempty"`
+	Nginxc                     string `yaml:"nginxc,omitempty"`
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/manifest/manifest.go:89]: Why do we have the `c` as the suffix? Can it be just `nginx` or `dynamicPluginProxy`?

File: pkg/apis/core/v1/storagecluster.go [hunk 4]
```diff
@@ -610,6 +610,8 @@ type ComponentImages struct {
 	KubeScheduler              string `json:"kubeScheduler,omitempty"`
 	KubeControllerManager      string `json:"kubeControllerManager,omitempty"`
 	Pause                      string `json:"pause,omitempty"`
+	DynamicPlugin              string `json:"dynamicPlugin,omitempty"`
+	Nginxc                     string `json:"nginxc,omitempty"`
```
Comment (piyush-nimbalkar) [inline pkg/apis/core/v1/storagecluster.go:614]: Same as above about the suffix.

File: deploy/crds/core_v1_storagecluster_crd.yaml [hunk 5]
```diff
@@ -3885,6 +3885,12 @@ spec:
                   pause:
                     type: string
                     description: Desired image for pause image.
+                  dynamicPlugin:
+                    type: string
+                    description: Desired image for dynamic plugin image.
+                  nginxinc:
+                    type: string
+                    description: Desired image for nginx-unprivileged image.
```
Comment (piyush-nimbalkar) [inline deploy/crds/core_v1_storagecluster_crd.yaml:3893]: Let's just say `Desired image for nginx proxy`. We don't have to mention what flavor we are using.

File: drivers/storage/portworx/component/plugin.go (hunk 1 above)
Comment (piyush-nimbalkar) [reply drivers/storage/portworx/component/plugin.go:365]: Additionally, the manifest should be adding the default values and returning them. Just follow the pattern that other component images use. Let's assume that the desired image is populated here.

**Review by piyush-nimbalkar — CHANGES_REQUESTED** [review]

File: drivers/storage/portworx/component/plugin.go (hunk 1 above)
Comment (nikita-bhatia) [reply drivers/storage/portworx/component/plugin.go]: done

File: drivers/storage/portworx/component/plugin.go (hunk 2 above)
Comment (nikita-bhatia) [reply drivers/storage/portworx/component/plugin.go]: done

File: drivers/storage/portworx/manifest/manifest.go (hunk 3 above)
Comment (nikita-bhatia) [reply drivers/storage/portworx/manifest/manifest.go]: changed to DynamicPluginProxy name

File: drivers/storage/portworx/manifest/manifest.go [hunk 6]
```diff
@@ -240,6 +246,14 @@ func fillDefaults(
 	if rel.Components.NodeWiper == "" {
 		rel.Components.NodeWiper = defaultNodeWiperImage
 	}
+	if rel.Components.DynamicPlugin == "" {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/manifest/manifest.go:249]: Can you add these after line 224 as well? That is used when we are not able to connect to the spec generator for some reason.

File: drivers/storage/portworx/manifest/manifest.go [hunk 7]
```diff
@@ -240,6 +246,14 @@ func fillDefaults(
...
+		rel.Components.DynamicPlugin = defaultDynamicPluginImage
+	}
+
+	if rel.Components.DynamicPluginProxy == "" {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/manifest/manifest.go:253]: Can you add UTs for these in manifest_test.go?

File: drivers/storage/portworx/portworx.go [hunk 8]
```diff
@@ -371,6 +371,14 @@ func (p *portworx) SetDefaultsOnStorageCluster(toUpdate *corev1.StorageCluster)
 			}
 		}
 
+		if toUpdate.Status.DesiredImages.DynamicPlugin == "" || pxVersionChanged {
```
Comment (piyush-nimbalkar) [inline drivers/storage/portworx/portworx.go:374]: Can you add UTs for this?

**Review by piyush-nimbalkar — CHANGES_REQUESTED** [review]

File: drivers/storage/portworx/manifest/manifest.go (hunk 6 above)
Comment (nikita-bhatia) [reply drivers/storage/portworx/manifest/manifest.go:251]: added

File: drivers/storage/portworx/manifest/manifest.go (hunk 7 above)
Comment (nikita-bhatia) [reply drivers/storage/portworx/manifest/manifest.go:255]: aded these components in manifest_test.go

File: drivers/storage/portworx/portworx.go (hunk 8 above)
Comment (nikita-bhatia) [reply drivers/storage/portworx/portworx.go:374]: added

**Review by piyush-nimbalkar — APPROVED** [review]

**Pull request merged**